core/services/workflows/v2/secrets.go

```diff
@@ -2,17 +2,244 @@ package v2
+	vaultCap, err := s.capRegistry.GetExecutable(ctx, vault.CapabilityID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get vault capability: %w", err)
+	}
+
+	vp := &vault.GetSecretsRequest{
+		Requests: make([]*vault.SecretRequest, 0),
+	}
+
+	logKeys := []string{}
+	for _, r := range request.Requests {
+		logKeys = append(logKeys, keyFor(s.workflowOwner, r.Namespace, r.Id))
+		vp.Requests = append(vp.Requests, &vault.SecretRequest{
+			Id:        r.Id,
+			Namespace: r.Namespace,
+			Owner:     s.workflowOwner,
+
+			// TODO: replace with actual encryption public key
+			EncryptionKeys: []string{"TODO_ENCRYPTION_PUBLIC_KEY"},
+		})
+	}
+
+	anypbReq, err := anypb.New(vp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert vault request to any: %w", err)
+	}
+
+	lggr := logger.With(s.lggr, "requestedKeys", logKeys, "owner", s.workflowOwner, "workflow", s.workflowName)
+	lggr.Debug("fetching secrets...")
+
+	resp, err := vaultCap.Execute(ctx, capabilities.CapabilityRequest{
+		Payload:      anypbReq,
+		Method:       vault.MethodGetSecrets,
+		CapabilityId: vault.CapabilityID,
+		Metadata: capabilities.RequestMetadata{
+			WorkflowOwner: s.workflowOwner,
+			WorkflowName:  s.workflowName,
+		},
+	})
+	if err != nil {
+		lggr.Errorw("failed to fetch secrets", "err", err)
+		return nil, fmt.Errorf("failed to execute vault.GetSecrets: %w", err)
+	}
+
+	lggr.Debug("successfully fetched secrets")
+
+	respPayload := &vault.GetSecretsResponse{}
+	err = resp.Payload.UnmarshalTo(respPayload)
+	if err != nil {
+		lggr.Errorw("failed to unmarshal vault payload to GetSecretsResponse", "err", err)
+		return nil, fmt.Errorf("failed to unmarshal vault payload to GetSecretsResponse: %w", err)
+	}
+
+	m := map[string]*vault.SecretResponse{}
+	for _, s := range respPayload.Responses {
+		key := keyFor(s.Owner, s.Namespace, s.Id)
+		m[key] = s
+	}
+
+	sdkResp := []*sdkpb.SecretResponse{}
+	for _, r := range request.Requests {
+		key := keyFor(s.workflowOwner, r.Namespace, r.Id)
+		resp, ok := m[key]
+		if !ok {
+			lggr.Debugw("could not find secret in response map", "key", key)
+			sdkResp = append(sdkResp, &sdkpb.SecretResponse{
+				Response: &sdkpb.SecretResponse_Error{
+					Error: &sdkpb.SecretError{
+						Id:        r.Id,
+						Namespace: r.Namespace,
+						Owner:     s.workflowOwner,
+						Error:     "could not find secret for " + key,
+					},
+				},
+			})
+			continue
+		}
+
+		if resp.Error != "" {
+			lggr.Debugw("secret request returned an error", "key", key, "err", resp.Error)
+			sdkResp = append(sdkResp, &sdkpb.SecretResponse{
+				Response: &sdkpb.SecretResponse_Error{
+					Error: &sdkpb.SecretError{
+						Id:        r.Id,
+						Namespace: r.Namespace,
+						Owner:     s.workflowOwner,
+						Error:     resp.Error,
+					},
+				},
+			})
+			continue
+		}
+
+		if len(resp.EncryptedDecryptionKeyShares) != 1 {
```

Comment:
Should this be more like this? I'm assuming that the engine will have a single key most of the time, but we should probably check against the length anyway.

```suggestion
		if len(resp.EncryptedDecryptionKeyShares) != len(s.EncryptionKeys) {
```

Btw, s.EncryptionKeys doesn't exist. I'm suggesting you hardcode it in the constructor or take a param there.

Reply:
Can I take this as a follow up @DeividasK ? I think this logic will likely need to change depending on the implementation of don2don